File: backend/src/main/java/com/codesmell/entity/Commit.java
package com.codesmell.entity;

import com.fasterxml.jackson.databind.JsonNode;
import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Type;

import java.time.LocalDateTime;

@Entity
@Table(name = "commits")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Commit {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "repository_id", nullable = false)
    private Repository repository;

    @Column(nullable = false)
    private String sha;

    private String message;

    private String author;

    private Integer additions;

    private Integer deletions;

    @Column(name = "files_changed")
    private Integer filesChanged;

    @Column(name = "committed_at")
    private LocalDateTime committedAt;

    @Type(JsonBinaryType.class)
    @Column(columnDefinition = "jsonb")
    private JsonNode features;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}

File: Code-Smell-Time-Machine-AI-Powered-Code-Risk-Prediction/app/models/commit.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON, func, UniqueConstraint
from app.database import Base


class Commit(Base):
    __tablename__ = "commits"

    id = Column(Integer, primary_key=True, index=True)
    repository_id = Column(Integer, ForeignKey("repositories.id", ondelete="CASCADE"), nullable=False, index=True)
    sha = Column(String, nullable=False)
    message = Column(String, nullable=False)
    author = Column(String, nullable=False)
    additions = Column(Integer, default=0)
    deletions = Column(Integer, default=0)
    files_changed = Column(Integer, default=0)
    committed_at = Column(DateTime, nullable=False)
    features = Column(JSON, nullable=True)
    created_at = Column(DateTime, server_default=func.now())

    __table_args__ = (UniqueConstraint("repository_id", "sha", name="unique_repo_sha"),)
